instructor_notes/fastq_subset/generate_fastq_subsets.sh

```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+set -euo pipefail
+
+# This script downloads and subsets FASTQ files which can be used during workshops when
+#  internet speed is limited and prohibits timely downloading of the full FASTQ files.
+
+
+# Function to curl, subset, and zip a FASTQ file
+# Provide a single argument: the FASTQ name, without `.gz` extension
+function subset_fastq() {
+    # Grab the file and unzip
+    curl -O https://sra-download.ncbi.nlm.nih.gov/traces/sra63/SRZ/011518/SRR11518889/$1.gz
+    # overwrite if already exists
+    gunzip -f $1.gz > $1
+
+    # Subset to first 10,000 reads (40000 lines)
+    echo "$(head -n 40000 $1)" > $1
```

Comment:
A couple things here I don't love... One is that the file name for the subsets is the same as the original, which seems like it has potential for confusion. The second is just that it seems like we should just be using pipes to make this more efficient and save disk space. 

For the first, I am adding the subset flag to the front so that it does not conflict with any later pattern matching, and for the latter, well, pipes. I _think_ this actually makes the whole thing faster, as curl only has to download as much as is needed for the head to complete, but that wasn't the main point

I am also including the gzipping here, because I see no reason to keep the unzipped version around. This also means we could make the input the full file name with `.gz`, but I left it as is.

```suggestion
    # Grab the file, unzip and take first 10000 records (40000 lines)
    curl -s https://sra-download.ncbi.nlm.nih.gov/traces/sra63/SRZ/011518/SRR11518889/$1.gz \
    | gunzip \
    | head -n 40000 \
    | gzip \
    > subset-${1}.gz

```

Reply:
I accepted this suggestion, but this code isn't working reliably for me... as in, when I run the script literally nothing happens, _sometimes_! Sometimes I get an empty R1 file, but no R2. Can you check how this runs on your end? SRA has been hanging a bit this morning, but I would expect that kind of problem to cause the script to drag on, not suddenly stop without actually doing anything.

Reply:
Took a second, but I forgot that piping `gzip` to `head` gives a broken pipe error. A potential "fix" is to change `set -euo pipefail` to just `set -u`.

The broken pipe error is pretty minimal, and the savings for not downloading the whole file may be worth it for this use case.

Reply:
Note also the following gzip line needs to be deleted, which did not happen in my suggestion.

Reply:
> Note also the following gzip line needs to be deleted, which did not happen in my suggestion.

Yep, I got this already.

I tried changing to `set -u`, but still nothing happens 🤷‍♀️ . Removed the line entirely, and still nothing. So, I ended up getting it working by having the `curl` as a separate step and then piping the rest. Adding back in `set -euo pipefail` works too with this. For some piping-related reason probably(?), the uncompressed fastq files are retained so I also added a line to remove those directly.